File: 206. Reverse Linked List/recursive_solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    var tail *ListNode
    newHead := reverse(head, tail)
    return newHead
}

func reverse(head *ListNode, tail *ListNode) (*ListNode) {
    if head == nil {
        return tail
    } else {
        node := head
        head = head.Next
        node.Next = tail
        tail = node
        return reverse(head, tail)
    }
}

File: 206. Reverse Linked List/recursive_solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tail = None
        return self.myRev(head, tail)
    
    def myRev(self, head, tail):
        if not head:
            return tail
        else:
            tmp = head.next
            head.next = tail
            return self.myRev(tmp, head)
            
            
